pkg/istio-agent/sds-agent.go

```diff
@@ -332,35 +333,26 @@ func newSecretCache(serverOptions sds.Options) (workloadSecretCache *cache.Secre
 		// If an explicit CA is configured, assume it is mounting /etc/certs
 		var rootCert []byte
 
-		// explicitSecret is true if a /etc/certs/root-cert file has been mounted. Will be used
-		// to authenticate the certificate of the SDS server (istiod or custom).
-		explicitSecret := false
-
-		if _, err := os.Stat(mountedRoot); err == nil {
-			rootCert, err = ioutil.ReadFile(mountedRoot)
-			if err != nil {
-				log.Warnf("Failed to load existing citadel root: %v", err)
-			} else {
-				explicitSecret = true
-			}
-		}
-
 		tls := true
 
 		if serverOptions.CAEndpoint == "" {
-			// Determine the default address, based on the presence of Citadel secrets
-			if explicitSecret {
-				log.Info("Using citadel CA for SDS")
-				serverOptions.CAEndpoint = "istio-citadel.istio-system:8060"
+			// When serverOptions.CAEndpoint is nil, the default CA endpoint
+			// will be a hardcoded default value (e.g., the namespace will be hardcoded
+			// as istio-system).
+			log.Info("Istio Agent uses default istiod CA")
+			serverOptions.CAEndpoint = "istio-pilot.istio-system.svc:15012"
+
+			if rootCert, err = ioutil.ReadFile(CitadelCACertPath + "/" + bootstrap.CACertNamespaceConfigMapDataName); err == nil {
+				log.Info("istiod uses self-issued certificate")
+			} else if rootCert, err = ioutil.ReadFile(k8sCAPath); err == nil {
+				log.Infof("istiod uses the k8s root certificate %v", k8sCAPath)
+			} else if rootCert, err = ioutil.ReadFile(cache.ExistingRootCertFile); err == nil {
```

Comment:
doesn't this need to be first? k8s root CA will always be there on kubernetes, so you could never have a custom CA?

Note that before this PR this was first as well

Reply:
As some k8s do not have signing APIs, Citadel issued istiod cert is put as first.

Reply:
So if you have k8s signing api (which is most people), it is impossible to have a custom root cert with this change. That does not seem to be what we want. Instead, I think the order should be `root cert override -> CitadelCACertPath -> k8s ca`, unless I am misunderstanding this

Reply:
The order has been revised to: root cert override -> CitadelCACertPath -> k8s ca.

Reply:
Per discussion, the original order remains because /etc/certs/root-cert.pem from Citadel is always mounted. A separate PR will address the problems.

Reply:
https://github.com/istio/istio/pull/20134